File: 2020/9/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input := strings.Split(string(data), "\n")
	parsedInput := []int{}
	for _, s := range input {
		n, err := strconv.Atoi(s)
		if err != nil {
			panic(err)
		}
		parsedInput = append(parsedInput, n)
	}
	invalidNum := part1(parsedInput)
	fmt.Println(invalidNum)
	fmt.Println(part2(parsedInput, invalidNum))
}

func part1(input []int) int {
	preambleLen := 25
	preamble := map[int]bool{}
	for i := 0; i < preambleLen; i++ {
		preamble[input[i]] = true
	}
	for i := preambleLen; i < len(input); i++ {
		valid := false
		for n := range preamble {
			if preamble[input[i]-n] && n != input[i]-n {
				valid = true
				break
			}
		}
		if !valid {
			return input[i]
		}
		preamble[input[i]] = true
		delete(preamble, input[i-preambleLen])
	}
	return -1
}

func part2(input []int, invalidNum int) int {
	start := 0
	finish := 1
	sum := input[0] + input[1]
	for sum != invalidNum {
		finish++
		sum += input[finish]
		for sum > invalidNum {
			sum -= input[start]
			start++
		}
	}
	contiguousRange := input[start : finish+1]
	sort.Ints(contiguousRange)
	return contiguousRange[0] + contiguousRange[len(contiguousRange)-1]
}

File: 2020/9/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map(Number);

function part1(input) {
  const preamble = new Set(input.slice(0, 25));

  function checkForSum(n, preamble) {
    for (const num of preamble) {
      if (preamble.has(n - num)) {
        return true;
      }
    }
    return false;
  }

  if (!checkForSum(input[25], preamble)) {
    return input[25];
  }

  for (let i = 26; i < input.length; i++) {
    preamble.delete(input[i - 26]);
    preamble.add(input[i - 1]);
    if (!checkForSum(input[i], preamble)) {
      return input[i];
    }
  }
}

const target = part1(input);
console.log(target);

let a = 0;
let b = 0;
let sum = input[0];
while (sum !== target) {
  if (sum < target) {
    // b++;
    // sum += input[b];
    sum += input[++b];
  } else {
    // sum -= input[a];
    // a++;
    sum -= input[a++];
  }
}

const rng = input.slice(a, b + 1);
console.log(Math.min(...rng) + Math.max(...rng));
